src/Models/Hooks/useTimeSeriesData.ts

```diff
@@ -0,0 +1,202 @@
+            adapter,
+            connectionString,
+            twins,
+            pollingInterval
+        ],
+        isAdapterCalledOnMount: false,
+        isLongPolling: pollingInterval ? true : false,
+        pollingIntervalMillis: pollingInterval
+    });
+
+    const onMountDate = useMemo(() => Date.now(), []);
+
+    /**
+     * After getting time series data from adapter, parse it in a structure of
+     * [timestamp, value] and store it as data to expose for consumers of this hook
+     *  */
+    useEffect(() => {
+        if (timeSeriesData?.adapterResult?.result?.data) {
+            logDebugConsole(
+                'debug',
+                `[END]: Fetching time series data finished: `,
+                timeSeriesData?.adapterResult?.result?.data
+            );
+            setData(timeSeriesData?.adapterResult?.result?.data);
+        } else {
+            setData(null);
+        }
+    }, [timeSeriesData?.adapterResult.result]);
+
+    useEffect(() => {
+        if (connectionString) {
+            setConnectionToQuery(
+                generateADXConnectionFromString(connectionString)
+            );
+        } else {
+            setConnectionToQuery(adapter?.getADXConnectionInformation());
+        }
+    }, [connectionString, adapter]);
+
+    const prevQuery = usePrevious(query);
+    useEffect(() => {
+        let newQuery = '';
+        if (connectionToQuery && quickTimeSpan && twins?.length) {
+            const timeFrom = new Date(onMountDate - quickTimeSpan);
+            const timeTo = new Date(onMountDate);
+            newQuery = getBulkADXQueryFromTimeSeriesTwins(
+                twins,
+                { from: timeFrom, to: timeTo },
+                connectionToQuery
+            );
+        }
+        if (prevQuery !== newQuery) {
+            setQuery(newQuery);
+        }
+    }, [connectionToQuery, quickTimeSpan, twins, onMountDate]);
+
+    const prevDeeplink = usePrevious(deeplink);
+    useEffect(() => {
+        const newDeeplink = getADXDeeplinkForWeb(connectionToQuery, query);
+        if (prevDeeplink !== newDeeplink) {
+            setDeeplink(newDeeplink);
+        }
+    }, [connectionToQuery, query]);
+
+    const fetchData = () => {
+        logDebugConsole(
+            'debug',
+            `[START]: Fetching time series data using query "${query}"`
+        );
+        timeSeriesData.callAdapter({
+            query,
+            connection: connectionToQuery
+        });
+    };
+
+    return useMemo(() => {
+        return {
+            query,
+            deeplink,
+            data,
+            fetchTimeSeriesData: fetchData,
+            isLoading: timeSeriesData.isLoading
+        };
+    }, [query, deeplink, data, fetchData, timeSeriesData.isLoading]);
+};
+
+const generateADXConnectionFromString = (
+    connectionString: string
+): IADXConnection | null => {
+    if (connectionString) {
+        try {
+            const parts = connectionString.split(';');
+            return {
```

Comment:
This is where we will want to harden our logic a bit. Since you are using these strings inside the query itself and this query string comes from the user's config file, this is an easy attack vector for inserting malicious stuff. Idk what they would do with that, but I still think it's worth thinking about how we can harden the logic here. 🤔

Reply:
There is isValidADXClusterUrl check before making any requests even they change the connection string into a malicious one in the config file. I will keep an eye on this one though.